pennylane_lightning/src/util/Util.hpp

```diff
@@ -199,87 +205,509 @@ template <class T> inline size_t dimSize(const std::vector<T> &data) {
+                               size_t m1, size_t m2, size_t n1, size_t n2) {
+    constexpr size_t BLOCK_THRESHOLD = 16;
+    size_t r, s, r1, s1, r2, s2;
+tail:
+    r1 = m2 - m1;
+    s1 = n2 - n1;
+    if (r1 >= s1 && r1 > BLOCK_THRESHOLD) {
+        r2 = (m1 + m2) / 2;
+        CFTranspose(mat, mat_t, m, n, m1, r2, n1, n2);
+        m1 = r2;
+        goto tail;
+    } else if (s1 > BLOCK_THRESHOLD) {
+        s2 = (n1 + n2) / 2;
+        CFTranspose(mat, mat_t, m, n, m1, m2, n1, s2);
+        n1 = s2;
+        goto tail;
+    } else {
+        for (r = m1; r < m2; ++r) {
+            for (s = n1; s < n2; ++s) {
+                mat_t[s * m + r] = mat[r * n + s];
+            }
+        }
+    }
+}
+
+/**
+ * @brief Transpose a matrix of shape m * n to n * m using the
+ * best available method.
+ *
+ * @tparam T Floating point precision type.
+ * @param mat Row-wise flatten matrix of shape m * n.
+ * @param m Number of rows of `mat`.
+ * @param n Number of columns of `mat`.
+ * @return mat transpose of shape n * m.
+ */
+template <class T>
+inline auto Transpose(const std::vector<std::complex<T>> mat, size_t m,
+                      size_t n) -> std::vector<std::complex<T>> {
+    if (mat.size() != m * n) {
+        throw std::invalid_argument("Invalid m & n for the input matrix");
+    }
+
+    std::vector<std::complex<T>> mat_t(n * m);
+    CFTranspose(mat.data(), mat_t.data(), m, n, 0, m, 0, n);
+    return mat_t;
+}
+
+/**
+ * @brief Calculates matrix-matrix product using OpenMP.
+ *
+ * @tparam T Floating point precision type.
+ * @param m_left Row-wise flatten matrix of size m * k.
+ * @param m_right Row-wise flatten matrix of size k * n.
+ * @param m_out Pre-allocated row-wise flatten matrix of size m * n.
+ * @param m Number of rows of `m_left`.
+ * @param n Number of columns of `m_right`.
+ * @param k Number of rows of `m_right`.
+ * @param transpose If `true`, requires transposed version of `m_right`.
+ *
+ * @note Consider transpose=true, to get a performance as close to CBLAS.
+ */
+template <class T>
+inline void omp_matrixMatProd(const std::complex<T> *m_left,
+                              const std::complex<T> *m_right,
+                              std::complex<T> *m_out, size_t m, size_t n,
+                              size_t k, bool transpose = false) {
+    if (!m_out) {
+        return;
+    }
+#if defined _OPENMP
+#pragma omp parallel default(none)
+#endif
+    {
+        std::size_t r, c, b;
+        if (transpose) {
+#if defined _OPENMP
+#pragma omp for
+#endif
+            for (r = 0; r < m; ++r) {
+                for (c = 0; c < n; ++c) {
+                    for (b = 0; b < k; ++b) {
+                        m_out[r * n + c] +=
+                            m_left[r * k + b] * m_right[c * n + b];
+                    }
+                }
+            }
+        } else {
+            std::size_t i, j, l;
+            std::size_t s = 2; // stride
```

Comment:
```suggestion
            std::size_t i, j, l;
            std::size_t stride = 2; 
```
As mentioned above, this is the way to go

Reply:
This is fixed!